vlasovsolver/cpu_trans_map_amr.cpp

```diff
@@ -334,103 +334,108 @@ bool trans_map_1d_amr(const dccrg::Dccrg<spatial_cell::SpatialCell,dccrg::Cartes
    int memsetTimerId = phiprof::initializeTimer("trans-amr-MemSet");
    int propagateTimerId = phiprof::initializeTimer("trans-amr-propagatePencil");
 
-#pragma omp parallel
+   const size_t blocksSize {unionOfBlocks.size()};
+   const size_t binsSize {DimensionPencils[dimension].activeBins.size()};
+
+   #pragma omp parallel
    {
+      phiprof::Timer mappingTimer {mappingTimerId}; // mapping (top-level)
+
       // Vector of pointers to cell block data, used for both reading and writing
       std::vector<Vec> blockDataBuffer(DimensionPencils[dimension].sumOfLengths*WID3/VECL);
       std::vector<Realf*> cellBlockData(DimensionPencils[dimension].sumOfLengths);
       std::vector<uint> pencilBlocksCount(DimensionPencils[dimension].N);
 
       // Loop over velocity space blocks (threaded).
-#pragma omp for schedule(dynamic,1)
-      for(uint blocki = 0; blocki < unionOfBlocks.size(); blocki++) {
-         // Get global id of the velocity block
-         vmesh::GlobalID blockGID = unionOfBlocks[blocki];
-
-         phiprof::Timer mappingTimer {mappingTimerId}; // mapping (top-level)
-
-         // Load data for pencils.
-         phiprof::Timer loadTimer {loadTimerId};
-         for (uint pencili = 0; pencili < DimensionPencils[dimension].N; ++pencili){
-            int nonEmptyBlocks = 0;
-            int L = DimensionPencils[dimension].lengthOfPencils[pencili];
-            int start = DimensionPencils[dimension].idsStart[pencili];
-            // Loop over cells in pencil
-            for (int b = 0; b < L; b++) {
-               // Get cell pointer and local block id
-               SpatialCell* srcCell = mpiGrid[DimensionPencils[dimension].ids[start + b]];
-               const vmesh::LocalID blockLID = srcCell->get_velocity_block_local_id(blockGID,popID);
-               // Store block data pointer for both loading of data and writing back to the cell
-               if (blockLID != srcCell->invalid_local_id()) {
-                  // Get data pointer
-                  cellBlockData[start + b] = srcCell->get_data(blockLID,popID);
-                  nonEmptyBlocks++;
-               } else {
-                  cellBlockData[start + b] = NULL;
+      // Get global id of the velocity block
+      // Load data for pencils.
+      #pragma omp for schedule(dynamic,1) collapse(2)
+      for(uint blocki = 0; blocki < blocksSize; blocki++) {
+         for (uint nBin = 0; nBin < binsSize; ++nBin) {
+            // For each block + bin we copy first copy each pencil's data into a buffer, clear the target blocks, and then sum the translated pencils in
+
+            phiprof::Timer loadTimer {loadTimerId};
+            vmesh::GlobalID blockGID = unionOfBlocks[blocki];
```

Comment:
Line 359 is a doozy to read. How about having 
`const int currentActiveBin = DimensionPencils[dimension].activeBins[nBin];` on the above line, then having the loop as 
`for (uint pencili : DimensionPencils[dimension].binsPencils[currentActiveBin]) {` please?